backend/pkg/rate/limit_http.go

```diff
@@ -98,88 +139,92 @@ func (h *HTTPLimiter) MatchHTTPPattern(
 	)
 	if groupTemplate != "" {
 		// Compile eventTemplate:
-		t, err = h.template.Clone()
+		t, err = h.rootTemplate.Clone()
 		if err == nil {
 			_, err = t.Parse(groupTemplate)
 		}
 		if err != nil {
 			return fmt.Errorf("error parsing group_template: %w", err)
 		}
 	}
-	limiterHandle := handle{
+	limiterMatcher := matcher{
 		HTTPLimiter:   h,
 		groupTemplate: t,
 	}
-	h.mux.Handle(pattern, limiterHandle)
+	h.httpMux.Handle(pattern, limiterMatcher)
 	return nil
 }
 
-type handle struct {
+// matcher is the HTTPHandle
+type matcher struct {
 	*HTTPLimiter
 	groupTemplate *template.Template
 }
 
-// ServeHTTP implements a basic http.Handler so that handler can be used
-// as a handler for the mux. It will only write on errors and is expected
-// to continue to the actual handler on success.
-func (h *HTTPLimiter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	ctx := r.Context()
+func (h *HTTPLimiter) handleRequest(r *http.Request) error {
 	if h.rewriteRequests {
 		r = rest.RewriteForwardedRequest(r)
 	}
 	res, err := h.Reserve(r)
 	if err != nil {
-		hdr := w.Header()
-		hdr.Set("Content-Type", "application/json")
-		w.WriteHeader(http.StatusInternalServerError)
-		_ = json.NewEncoder(w).Encode(rest.Error{
-			Err:       "internal server error",
-			RequestID: requestid.FromContext(ctx),
-		})
-		return
+		return err
 	}
 	if res == nil || res.OK() {
-		return
+		return nil
 	} else {
-		hdr := w.Header()
-		hdr.Set("Content-Type", "application/json")
-		retryAfter := math.Ceil(res.Delay().Abs().Seconds())
-		hdr.Set("Retry-After", fmt.Sprint(retryAfter))
-		w.WriteHeader(http.StatusTooManyRequests)
-		_ = json.NewEncoder(w).Encode(rest.Error{
-			Err:       ErrTooManyRequests.Error(),
-			RequestID: requestid.FromContext(ctx),
-		})
+		return &TooManyRequestsError{
+			Delay: res.Delay(),
+		}
 	}
 }
 
-func (h *HTTPLimiter) MiddlewareGin(c *gin.Context) {
-	r := c.Request
-	if h.rewriteRequests {
-		r = rest.RewriteForwardedRequest(r)
-	}
-	res, err := h.Reserve(r)
+func handleError(w http.ResponseWriter, err error) {
+	var tooManyRequests *TooManyRequestsError
+	status := http.StatusInternalServerError
+	hdr := w.Header()
+	hdr.Set("Content-Type", "application/json")
```

Comment:
I just wondered now, do we return any json here?

Reply:
Ohh man, all this refactoring made me remove the part writing the body :exploding_head: